### ISSUE TITLE
[Bug] Peers disconnect upon requesting 20 blocks

### ISSUE THREAD

**Issue by CumpsD** [iron-fish/ironfish#725]
# TL;DR

## Running from source (Linux)

```bash
sed -i 's%const REQUEST_BLOCKS_PER_MESSAGE.*%const REQUEST_BLOCKS_PER_MESSAGE = 2%' ironfish/src/syncer.ts
yarn start:once start
```

## Running from source (Windows)

* Open ironfish/src/syncer.ts in Notepad
* Edit `const REQUEST_BLOCKS_PER_MESSAGE = 20` to `const REQUEST_BLOCKS_PER_MESSAGE = 2` 
* Save the file
* `yarn start:once start`

## Running from Docker (Linux and Windows)

**❗This assumes you did NOT run your container with `--rm` else the changes are gone on restart!!❗**

I expect your container to be run with:
`docker run -d --name ironfish --network host ghcr.io/iron-fish/ironfish:latest ........`

Find NAME_OF_YOUR_DOCKER_CONTAINER with `docker ps -a`

![image](https://user-images.githubusercontent.com/1201644/146088820-6f730840-20ad-456a-83e2-d68c91e13b85.png)

```bash
docker exec -t NAME_OF_YOUR_DOCKER_CONTAINER sh -c "sed -i 's%REQUEST_BLOCKS_PER_MESSAGE.*%REQUEST_BLOCKS_PER_MESSAGE = 2%' /usr/src/app/node_modules/ironfish/build/src/syncer.js"
docker restart NAME_OF_YOUR_DOCKER_CONTAINER
docker logs -f NAME_OF_YOUR_DOCKER_CONTAINER
```

------

# Problem

A lot of issues on GH and Discord about peers disconnecting while syncing. I've done some debugging:

Reproduction of the bug:
![image](https://user-images.githubusercontent.com/1201644/145732921-c0c68338-2607-48c8-aefc-a439320e7440.png)

My initial start:
https://github.com/iron-fish/ironfish/blob/master/ironfish/src/syncer.ts#L358-L362
```
      this.logger.info(
        `Requesting ${this.blocksPerMessage} blocks starting at ${HashUtils.renderHash(
          head,
        )} (${sequence}) from ${peer.displayName}`,
      )
```

I noticed blocksPerMessage is `20`, being passed in the ctor: https://github.com/iron-fish/ironfish/blob/master/ironfish/src/syncer.ts#L65

My hunch was the peers are timing out sending 20 blocks, so I lowered it to 1:

![image](https://user-images.githubusercontent.com/1201644/145732948-5a3bd685-ed0c-448d-8ffd-f50ac72985bb.png)

Upon restarting, this solved the problem and my chain is syncing again:

![image](https://user-images.githubusercontent.com/1201644/145732961-ebb8cf52-55d7-4ba5-a038-12eb36ea165f.png)

Experimenting with a number, I found 10 is still to high, but 5 works.

```
    this.blocksPerMessage = 5; //options.blocksPerMessage ?? REQUEST_BLOCKS_PER_MESSAGE
```

![image](https://user-images.githubusercontent.com/1201644/145732988-29f34213-999e-43fd-8622-b1fced0715e7.png)

Short term fix is to use a lower `blocksPerMessage`, long term fix is to make sure peers dont disconnect when sending blocks over.

I can create the PR for the short term fix if needed.

**Comment by sl0n** [comment]
In my case i set this.blocksPerMessage = 3, and then errors disappeared, and my node could proceed with syncing.

**Comment by CumpsD** [comment]
Yes, upon reaching the tip of the chain, 5 started giving errors too here. I changed it to 2 and am now back fully synced and mining.

**Comment by Enissay** [comment]
Thanks, it got stuck around 14k then it is now working with `this.blocksPerMessage = 10;`...

**Comment by Enissay** [comment]
OKay, I jynxed it... It got stuck at 16k, now down to `2`... Fingers crossed it finishes fine

**Comment by m4steel** [comment]
Thanks bro! I was breaking my head and you helped a lot!!!

**Comment by m4steel** [comment]
> const SYNCER_TICK_MS = 4 * 1000 const LINEAR_ANCESTOR_SEARCH = 3 const REQUEST_BLOCKS_PER_MESSAGE = 20
> 
> Just a question, the 20 here i just ignore???

You don't need to change that line, look at the picture man

**Comment by smkcoder02** [comment]
in mac, where is the file located? I have this file sitting under the path "System/Volumes/Data/usr/local/Cellar/ironfish/23/libexec/node_modules/ironfish/src/syncer.ts"

changed this file to 5 Blocks per message, then stopped the node and started it again but the log still says it is looking for 20 blocks. Any suggestions on this will be helpful. Node stopped syncing as below,

![Screen Shot 2021-12-13 at 1 21 20 AM](https://user-images.githubusercontent.com/96043360/145769197-e9e5a5c3-8873-48b1-b89f-d26e54260cb3.png)

**Comment by srwhitney19** [comment]
I also changed it to 5 in syncer.ts but re-running the node still shows requesting 20. any idea what I'm doing wrong? (ubuntu, downloaded from source)

**Comment by Enissay** [comment]
> I also changed it to 5 in syncer.ts but re-running the node still shows requesting 20. any idea what I'm doing wrong? (ubuntu, downloaded from source)

As soon as I was saving the file, the node was reloading automatically...
Make sure you modified the right line in the right file!

**Comment by srwhitney19** [comment]
i stopped the node from syncing,
edited the file at ./ironfish/ironfish/src/syncer.ts in the exact line
detailed above and its still requesting 20 blocks

any ideas anyone?

**Comment by Enissay** [comment]
Frankly I don't know... I just checked again and it is `line 65`

**Comment by shandongdacongge** [comment]
> the peers are timing out sending

u should 
`yarn build `
the code

**Comment by iephesus** [comment]
i use docker image and how to do this?

**Comment by novidi** [comment]
> 

i do all you wrote for docer but got this messege after reset node:SyntaxError: Invalid left-hand side in assignmen
![aa](https://user-images.githubusercontent.com/93491527/146096065-51ec06ce-8905-4cee-bab6-b7798188e89b.jpg)
t

**Comment by generalleger** [comment]
@novidi try this instead
```docker exec -t ironfish sh -c "sed -i 's%REQUEST_BLOCKS_PER_MESSAGE = 20.*%REQUEST_BLOCKS_PER_MESSAGE = 2%' /usr/src/app/node_modules/ironfish/build/src/syncer.js"```

**Comment by dguenther** [comment]
Thanks for the detailed bug reports and reproduction steps on this. A fix for this is available in 0.1.11, so please update your nodes and try again, and if you're still having further problems related to this, please open a new issue.
